extras/ai-battle/HeadlessGame.cpp

```diff
@@ -0,0 +1,264 @@
+
+    printConsole("┌───────────────┬───────────────────────┬───────────────────────┬────────────────┐\n");
+    printConsole(
+      "│ GF %10s │ Game Clock  %s │ Wall Clock  %s │ %7s GF/sec │\n", HumanReadableNumber(em_.GetCurrentGF()).c_str(),
+      ToString(SPEED_GF_LENGTHS[GameSpeed::Normal] * em_.GetCurrentGF()).c_str(), // elapsed time
+      ToString(std::chrono::duration_cast<std::chrono::milliseconds>(std::chrono::steady_clock::now() - gameStartTime_))
+        .c_str(),                                                       // wall clock
+      HumanReadableNumber(em_.GetCurrentGF() - lastReportGf_).c_str()); // GF per second
+    printConsole("└───────────────┴───────────────────────┴───────────────────────┴────────────────┘\n");
+    printConsole("\n");
+    printConsole("┌────────────────────────┬─────────────────┬─────────────┬───────────┬───────────┐\n");
+    printConsole("│ Player                 │ Country         │ Buildings   │ Military  │ Gold      │\n");
+    printConsole("├────────────────────────┼─────────────────┼─────────────┼───────────┼───────────┤\n");
+    for(unsigned playerId = 0; playerId < world_.GetNumPlayers(); ++playerId)
+    {
+        const GamePlayer& player = world_.GetPlayer(playerId);
+        printConsole("│ %s%-22s%s │ %15s │ %11s │ %9s │ %9s │\n", player.IsDefeated() ? "\x1b[9m" : "",
+                     player.name.c_str(), player.IsDefeated() ? "\x1b[29m" : "",
+                     HumanReadableNumber(player.GetStatisticCurrentValue(StatisticType::Country)).c_str(),
+                     HumanReadableNumber(player.GetStatisticCurrentValue(StatisticType::Buildings)).c_str(),
+                     HumanReadableNumber(player.GetStatisticCurrentValue(StatisticType::Military)).c_str(),
+                     HumanReadableNumber(player.GetStatisticCurrentValue(StatisticType::Gold)).c_str());
+    }
+    printConsole("└────────────────────────┴─────────────────┴─────────────┴───────────┴───────────┘\n");
+
+    lastReportGf_ = em_.GetCurrentGF();
+}
+
+std::vector<PlayerInfo> GeneratePlayerInfo(const std::vector<AI::Info>& ais)
+{
+    std::vector<PlayerInfo> ret;
+    for(const AI::Info& ai : ais)
+    {
+        PlayerInfo pi;
+        pi.ps = PlayerState::Occupied;
+        pi.aiInfo = ai;
+        switch(ai.type)
+        {
+            case AI::Type::Default: pi.name = "AIJH " + std::to_string(ret.size()); break;
+            case AI::Type::Dummy:
+            default: pi.name = "Dummy " + std::to_string(ret.size()); break;
+        }
+        pi.nation = Nation::Romans;
+        pi.team = Team::None;
+        ret.push_back(pi);
+    }
+    return ret;
+}
+
+#ifdef WIN32
+HANDLE setupStdOut()
+{
+    HANDLE h = GetStdHandle(STD_OUTPUT_HANDLE);
+    SetConsoleMode(h, ENABLE_PROCESSED_OUTPUT | ENABLE_VIRTUAL_TERMINAL_PROCESSING);
+    SetConsoleOutputCP(65001);
+    return h;
+}
+#endif
+
+void printConsole(const char* fmt, ...)
+{
+    static char buffer[512];
```

Comment:
Any reason to have this static?

```suggestion
    char buffer[512];
```

Reply:
done